### PR TITLE
Federation: Use equivalent() instead of == to compare DNS ResourceRecordSets. Fixes #28135

### LINKED ISSUE #28135: federation service controller has trouble deciding when there is no delta
E.g.,

```
I0627 22:26:46.221856       1 dns.go:215] Existing recordset                <(clouddns) "federated-service.e2e-tests-service-cuza5.federation.svc.us-central1-c.us-central1.kube.5yetis.net." type=CNAME rrdatas=["federated-service.e2e-tests-service-cuza5.federation.svc.us-central1.kube.5yetis.net."] ttl=180>
I0627 22:26:46.221885       1 dns.go:216] ... not equal to needed recordset <(clouddns) "federated-service.e2e-tests-service-cuza5.federation.svc.us-central1-c.us-central1.kube.5yetis.net." type=CNAME rrdatas=["federated-service.e2e-tests-service-cuza5.federation.svc.us-central1.kube.5yetis.net."] ttl=180>
I0627 22:26:46.221919       1 dns.go:217] ... removing existing and adding needed.
```

See also #28134

### PR DESCRIPTION
**Pull request by ghost**
Without this we sometimes incorrectly assume that DNS records hosted on cloud DNS servers need to be replaced when they don't need to be.  This results in unnecessary churn on DNS records.

cc: @nikhiljindal @madhusudancs @mfanjie FYI

### CODE REVIEW

File: federation/pkg/federation-controller/service/dns.go [hunk 1]
```diff
@@ -149,6 +150,24 @@ func getResolvedEndpoints(endpoints []string) ([]string, error) {
 	return resolvedEndpoints, nil
 }
 
+/* equavalent compares two ResourceRecordSets for semantic equivalence.
```
Comment (mml) [inline federation/pkg/federation-controller/service/dns.go]: typo

**Comment by mml** [comment]
Test?

**Comment by quinton-hoole** [comment]
@mml Thanks for keeping me honest.  I'll add some unit tests.

File: federation/pkg/federation-controller/service/dns.go [hunk 2]
```diff
@@ -149,6 +150,24 @@ func getResolvedEndpoints(endpoints []string) ([]string, error) {
...
+   Go's equality operator doesn't work the way we want it to in this case.
+   More specifically (from the Go spec):
+   "Two struct values are equal if their corresponding non-blank fields are equal."
+   In our case, there may be some private internal member variables that are not equal,
+   but we want the two structs to be considered equivalent anyway, if the fields exposed
+   via their interfaces are equal.
```
Comment (erictune) [inline federation/pkg/federation-controller/service/dns.go]: File a bug to remember to reflect to make sure new fields don't start appearing which make things which are different appear the same to this routing.

**Comment by matchstick** [comment]
We should resolve this in 1.3.x or 1.4. Moving to P0 and next-candidate so it does not get lost.
@quinton-hoole @mml @kubernetes/sig-cluster-federation

**Comment by justinsb** [comment]
Should we also replace this check with `equivalent`:

https://github.com/kubernetes/kubernetes/pull/28182/files#diff-d0d23fdd8ea45a98fb0d06ba3a44a7e8L208

(The tests actually caught this one, BTW, once I made some other fixes in https://github.com/kubernetes/kubernetes/pull/28668)

**Comment by quinton-hoole** [comment]
Yes, you're right @justinsb.  I'll change that one too.  Good to know that our tests are catching stuff :-)

**Comment by quinton-hoole** [comment]
@mml @justinsb MY apologies for the delay in finishing this off.  Done now.  PTAL.

File: federation/pkg/dnsprovider/dns_test.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+const testDNSZone string = "foo.com"
+
+var testData = []struct {
+	inputs         [2]record
+	expectedOutput bool
+}{
+	{
+		[2]record{
+			{"foo", []string{"1.2.3.4", "5,6,7,8"}, 180, "A"}, // Identical
```
Comment (justinsb) [inline federation/pkg/dnsprovider/dns_test.go]: nit: Commas are kinda weird here, right?

also might be nice to have a test to show that order matters in rrdata (which it should, I believe)

File: federation/pkg/federation-controller/service/dns.go [hunk 4]
```diff
@@ -205,7 +205,7 @@ func (s *ServiceController) ensureDnsRrsets(dnsZoneName, dnsName string, endpoin
 			// Need an appropriate CNAME record.  Check that we have it.
 			newRrset := rrsets.New(dnsName, []string{uplevelCname}, minDnsTtl, rrstype.CNAME)
 			glog.V(4).Infof("No healthy endpoints for %s.  Have recordset %v. Need recordset %v", dnsName, rrset, newRrset)
-			if rrset == newRrset {
+			if dnsprovider.ResourceRecordSetsEquivalent(rrset, newRrset) {
 				// The existing rrset is equal to the required one - our work is done here
 				glog.V(4).Infof("Existing recordset %v is equal to needed recordset %v, our work is done here.", rrset, newRrset)
```
Comment (justinsb) [inline federation/pkg/federation-controller/service/dns.go]: super nit: if you're going to change the other message from equal to equivalent ...

**Comment by justinsb** [comment]
Some very minor nits, but I would love to see this go in (even as-is; my comments are just nits) - LGTM and this is actually more of an issue on AWS where concurrent changes are blocked (as well as for a dependent PR I have).

File: federation/pkg/dnsprovider/dns_test.go (hunk 3 above)
Comment (quinton-hoole) [inline federation/pkg/dnsprovider/dns_test.go]: Good point.  Will do now.

File: federation/pkg/dnsprovider/dns_test.go [hunk 5]
```diff
@@ -0,0 +1,91 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
```
Comment (mml) [inline federation/pkg/dnsprovider/dns_test.go]: You need to remove "All rights reserved" to pass verification.

File: federation/pkg/dnsprovider/dns_test.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+	"testing"
+
+	"k8s.io/kubernetes/federation/pkg/dnsprovider/rrstype"
+)
+
+// Compile time interface check
+var _ ResourceRecordSet = record{}
+
+type record struct {
+	Name_    string
```
Comment (mml) [inline federation/pkg/dnsprovider/dns_test.go]: trailing underscore style... i haven't seen this elsewhere.  any reason to not just make these lower case?

**Comment by mml** [comment]
Test coverage for `ResourceRecordSetsEquivalent` is good, but I also think we should have a `kube-dns` test (unit if we can, but integration or e2e if we must) that when the desired rrset value matches the existing rrset value, we do exactly nothing.  Totally OK leaving this as a TODO comment at this point.

**Comment by quinton-hoole** [comment]
Thanks @mml, @justinsb.
I added the unit test @justinsb suggested, and appeased his other pedantry :-)

@mml, I agree, more unit testing of the of the federated service controller (I assume that's what you mean, even though you mention a kube-dns test?).  I'll add that in a separate PR, as you suggest.

This one LGTM?

Q

**Comment by quinton-hoole** [comment]
Sorry, I just noticed @mml's other comments.  Sending in another commit on this PR to address those.

**Comment by quinton-hoole** [comment]
OK, all done now, I think.  Ready for LGTM.

**Comment by mml** [comment]
LGTM

**Comment by justinsb** [comment]
@k8s-bot test this issue: #29355

**Comment by justinsb** [comment]
(I think that's the problem; I ran  hack/verify-munge-docs.sh  manually and it didn't seem to have a problem)

**Pull request merged**